cli/src/semgrep/core_runner.py

```diff
@@ -235,35 +261,121 @@ async def _core_stderr_processor(
             stderr=asyncio.subprocess.PIPE,
-            limit=1024 * 1024 * 1024,  # buffer limit to read in bytes
+            limit=INPUT_BUFFER_LIMIT,
             preexec_fn=setrlimits_preexec_fn,
         )
 
-        # Raise any exceptions from processing stdout/err
-        results = await asyncio.gather(
-            self._core_stdout_processor(process.stdout),
-            self._core_stderr_processor(process.stderr),
-            return_exceptions=True,
-        )
-        for r in results:
-            if isinstance(r, Exception):
-                raise SemgrepError(f"Error while running rules: {r}")
+        # Ensured by passing stdout/err named parameters above.
+        assert process.stdout
+        assert process.stderr
+
+        await self._handle_process_outputs(process.stdout, process.stderr)
 
         # Return exit code of cmd. process should already be done
         return await process.wait()
 
+    def _run_forked_analysis_in_child(self) -> None:
+        """
+        Run semgrep_analyze from within the child process.
+        """
+        # Like in the exec case, try to expand limits in the child.
+        setrlimits_preexec_fn()
+
+        # TYPE_CHECKING is always false at run time, but by doing this,
+        # mypy will know which module 'semgrep_bridge_python' refers to,
+        # and hence check the calls to it.
+        if TYPE_CHECKING:
+            import semgrep_bridge_python
+        else:
+            semgrep_bridge_python = SemgrepCore.get_bridge_module()
+
+        # Currently, we delay initializing the OCaml runtime until we
+        # are in the forked child.  Later, we may want to hoist this to
+        # the parent somewhere so multiple queries can be more
+        # efficiently performed in the context of a Snowflake UDF.
+        semgrep_bridge_python.startup()
+
+        # Invoke the semgrep_bridge_python module.
+        err = semgrep_bridge_python.semgrep_analyze(self._cmd, self._handle_read_file)
+
+        if err != None:
+            # Convey an error back to the parent.
+            print(err, file=sys.stderr)
+            sys.exit(2)
+
+        semgrep_bridge_python.shutdown()
+
+    async def _stream_fork_subprocess(self) -> int:
+        """
+        Run semgrep_bridge_python.so in a forked (but not exec'd) child
+        process, consuming its output asynchronously.
+
+        Return its exit code when it terminates.
+        """
+        process = await fork_subprocess.start_fork_subprocess(
+            lambda: self._run_forked_analysis_in_child(), limit=INPUT_BUFFER_LIMIT
+        )
+
+        await self._handle_process_outputs(process.stdout, process.stderr)
+
+        return process.wait()
+
     def execute(self) -> int:
         """
         Run semgrep-core and listen to stdout to update
         progress_bar as necessary
 
         Blocks til completion and returns exit code
         """
+        open_and_ignore("/tmp/core-runner-semgrep-BEGIN")
```

Comment:
what are these for?

Reply:
My comment on `open_and_ignore` was meant to explain this:

> Attempt to open 'fname' simply so a record of having done so will be seen by 'strace'.

I was using `strace` to test which files were opened by `semgrep-core`.  To do that, I wanted a marker in the `strace` output of where `semgrep-core` starts and ends in order to easily distinguish its calls from those of the Python CLI.